pyccel/ast/headers.py

```diff
@@ -734,6 +738,41 @@ def apply(self, args, results=None):
             newargs[i] = new
         return newargs
 
+    def make_necessary_copies(self, args, results):
+        """ Copy any arguments which are modified in fortran but
+        not in python into the result variable. If the variable
+        is modified in fortran then the argument and result
+        variable should be the same when apply is called
+
+        Parameters
+        ----------
+        args    : list of Variables
+                   The arguments passed to the macro function
+        results : list of Variables
+                  The results collected from the macro function
```

Comment:
You are right this is not clear. Let me try and explain what is happening and see if we can find a better description.

If we look at the example from the issue:
```
lu, piv, info = dgetrf(A)
```
The input to the macro in the code is: `A`
The results collected from the macro in the code are: `lu`, `piv`, `info`

The function `make_necessary_copies` should return:
final_args : `lu`
expr : `lu[:] = A[:]`

So that when `apply` is called it sees an intermediate macro containing:
```
lu, piv, info = dgetrf(lu)
```
At this stage:
The input to the macro (as defined in lapack.pyh) is: `lu`
The results collected from the macro (as defined in lapack.pyh) are: `lu`, `piv`, `info`

The macro (as defined in lapack.pyh) is then converted to the function call:
```
call dgetrf(size(lu, 2_C_INT64_T, C_INT64_T), size(lu, 1_C_INT64_T, &
      C_INT64_T), lu, size(lu, 2_C_INT64_T, C_INT64_T), piv, info)
```

I am not sure what you mean by "the input function"

Reply:
By input function I meant this part of the macro `(ab, IPIV, info), dgetrf(ab)`
And I thought you meant by a macro function, the representation of the function in fortran, this part of the macro `dgetrf(ab.shape[0], ab.shape[1], ab, ab.shape[0], IPIV, info)`,
Do you think replacing "macro function" by "macro" is more appropriate ?

Reply:
I disagree that this is a macro: `dgetrf(ab.shape[0], ab.shape[1], ab, ab.shape[0], IPIV, info)`
If in C you do:
```
#define GRAVITY 9.81
```
then `9.81` is not the macro. I would say that `GRAVITY` is the macro
Here we have:
```
#$ header macro (ab, IPIV, info), dgetrf(ab) := dgetrf(ab.shape[0], ab.shape[1], ab, ab.shape[0], IPIV, info)
```
so `(ab, IPIV, info), dgetrf(ab)` is the macro, and `dgetrf(ab.shape[0], ab.shape[1], ab, ab.shape[0], IPIV, info)` is what the macro is evaluated to (value of the macro? function associated with the macro?)

Maybe we could use this terminology:
- **macro definition** : `(ab, IPIV, info), dgetrf(ab)`
- **macro code usage** : `lu, piv, info = dgetrf(A)`
- **macro expansion**: `dgetrf(ab.shape[0], ab.shape[1], ab, ab.shape[0], IPIV, info)`

The results passed to this function are not collected from the macro definition, but from what is written in the code.

What do you think of the following modification?
```
Copy any arguments provided in python which the macro
definition indicates should match the results into the
corresponding result.

        Parameters
        ----------
        args    : list of Variables
                   The arguments passed to the macro in the python code
        results : list of Variables
                  The results collected from the macro in the python code

        Results
        -------
        final_args : list of Variables
                     The arguments which can be passed to the apply function
                     which match the expectations of the macro definition
        expr       : list of Assigns
                      Any Assigns necessary before the function (result of the 
                      macro expansion) returned by the apply function is called
```